pkg/controller/kmeshsecurity/security_client.go

```diff
@@ -0,0 +1,199 @@
+ */
+
+package kmeshsecurity
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/types/known/structpb"
+	pb "istio.io/api/security/v1alpha1"
+	"istio.io/istio/pkg/security"
+	"istio.io/istio/pkg/spiffe"
+	nodeagentutil "istio.io/istio/security/pkg/nodeagent/util"
+	pkiutil "istio.io/istio/security/pkg/pki/util"
+	"kmesh.net/kmesh/pkg/controller/workload"
+	"kmesh.net/kmesh/pkg/nets"
+)
+ 
+ var tlsOpts *TLSOptions 
+
+ type CitadelClient struct {
+	 // It means enable tls connection to Citadel if this is not nil.
+	 tlsOpts  *TLSOptions
+	 client   pb.IstioCertificateServiceClient
+	 conn     *grpc.ClientConn
+	 opts     *security.Options
+ }
+
+ type TLSOptions struct {
+	RootCert string
+	Key      string
+	Cert     string
+}
+ 
+ // NewCitadelClient create a CA client for Citadel.
+ func NewCitadelClient(opts *security.Options, tlsOpts *TLSOptions) (*CitadelClient, error) {
+	 var err error;
+ 
+	 c := &CitadelClient{
+		 tlsOpts:  tlsOpts,
+		 opts:     opts,
+	 }
+
+	 conn, err := nets.GrpcConnect(CSRSignAddress);
+	 if err != nil {
+		 log.Errorf("Failed to connect to endpoint %s: %v", opts.CAEndpoint, err)
+		 return nil, fmt.Errorf("failed to connect to endpoint %s", opts.CAEndpoint)
+	 }
+ 
+	 c.conn = conn
+	 c.client = pb.NewIstioCertificateServiceClient(conn)
+	 return c, nil
+ }
+
+// CSRSign calls Citadel to sign a CSR.
+func (c *CitadelClient) CSRSign(csrPEM []byte, certValidTTLInSec int64) (res []string, err error) {
+	crMetaStruct := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			security.CertSigner: {
+				Kind: &structpb.Value_StringValue{StringValue: c.opts.CertSigner},
+			},
+		},
+	}
+	req := &pb.IstioCertificateRequest{
+		Csr:              string(csrPEM),
+		ValidityDuration: certValidTTLInSec,
+		Metadata:         crMetaStruct,
+	}
+
+	defer func() {
+		if err != nil {
+			log.Errorf("failed to sign CSR: %v", err)
+			if err := c.reconnect(); err != nil {
+				log.Errorf("failed reconnect: %v", err)
+			}
+		}
+	}()
+
+	ctx := metadata.NewOutgoingContext(context.Background(), metadata.Pairs("ClusterID", c.opts.ClusterID))
+	resp, err := c.client.CreateCertificate(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("create certificate: %v", err)
+	}
+
+	if len(resp.CertChain) <= 1 {
+		return nil, errors.New("invalid empty CertChain")
+	}
+
+	return resp.CertChain, nil
+}
+
+// concatCerts concatenates PEM certificates, making sure each one starts on a new line
+func concatCerts(certsPEM []string) []byte {
+	if len(certsPEM) == 0 {
+		return []byte{}
+	}
+	var certChain bytes.Buffer
+	for i, c := range certsPEM {
+		certChain.WriteString(c)
+		if i < len(certsPEM)-1 && !strings.HasSuffix(c, "\n") {
+			certChain.WriteString("\n")
+		}
+	}
+	return certChain.Bytes()
+}
+
+func (c *CitadelClient) fetch_cert(uid string) (secret *security.SecretItem, err error) {
```

Comment:
Use camel case in golang
```suggestion
func (c *CitadelClient) fetchCert(uid string) (secret *security.SecretItem, err error) {
```